### ISSUE TITLE
Differentiable data augmentation

### ISSUE THREAD

**Issue by Rubiel1** [junyanz/pytorch-CycleGAN-and-pix2pix#1081]
There is a new paper on differentiable data augmentation for GANS: https://arxiv.org/pdf/2006.10738.pdf, it seems promising for GANS.

I have implemented it on pix2pix. For any one interested, I added: 

the [file](https://github.com/NewSciTLH/data-efficient-gans/blob/master/DiffAugment_pytorch.py) to the folder utils.

`parser.add_argument('--dif_data_augm', type=str, default='', help='differential data augmentation currenly only for pix2pix models. [color | translation | cutout | color,cutout|...|color,translation,cutout]')`
 [here](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/ebf2c791e7c94ff5ab72b7a35ad97135af275b95/options/base_options.py#L56) , 

`from util.DiffAugment_pytorch import DiffAugment` 
[here](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/ebf2c791e7c94ff5ab72b7a35ad97135af275b95/models/pix2pix_model.py#L4)

`# specify the smooth data augmentation, default is none,  options include translation, change on color and random cutout
        self.dif_data_augm = opt.dif_data_augm` 
[here](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/ebf2c791e7c94ff5ab72b7a35ad97135af275b95/models/pix2pix_model.py#L46)

` # smooth data aumentation policy
        policy = self.dif_data_augm` 
[here](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/ebf2c791e7c94ff5ab72b7a35ad97135af275b95/models/pix2pix_model.py#L92)

I replaced [this](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/ebf2c791e7c94ff5ab72b7a35ad97135af275b95/models/pix2pix_model.py#L94) by 
`fake_AB = DiffAugment(fake_AB.detach(), policy=policy)  # by default we don't modify the data
        pred_fake = self.netD(fake_AB)
        self.loss_D_fake = self.criterionGAN(pred_fake, False)`

and [this](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/ebf2c791e7c94ff5ab72b7a35ad97135af275b95/models/pix2pix_model.py#L98) by 
`        real_AB = DiffAugment(real_AB, policy=policy)
        pred_real = self.netD(real_AB)  # by default we don't modify the data`

**Comment by junyanz** [comment]
Great. Does it help in your experiments?
If so, maybe we can create a PR or a branch. @zsyzzsoft

**Comment by Rubiel1** [comment]
I was missing one line, so I updated my original comment. I found that this helps to prevent the discriminator to overfit. But I was missing one line so I have to repeat the experiments.
